drivers: can: native_posix_linux: add embedded libc support

henrikbrixandersen opened the pull request:
Add support for compiling the native POSIX Linux (SocketCAN) driver with an embedded C-library. Add build-only test for the native POSIX Linux (SocketCAN) driver.

henrikbrixandersen committed:
drivers: can: native_posix_linux: add embedded libc support

Add support for compiling the native POSIX Linux (SocketCAN) driver with an
embedded C-library.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
tests: drivers: build_all: can: add build test for native posix linux

Add build-only test for the native POSIX Linux (SocketCAN) driver.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

aescolar commented on drivers/can/CMakeLists.txt:
```diff
@@ -20,28 +20,30 @@ zephyr_library_sources_ifdef(CONFIG_CAN_STM32H7_FDCAN can_stm32h7_fdcan.c)
-      can_native_posix_linux_socketcan.c
-    )
+    zephyr_library_sources(can_native_posix_linux.c)
+    set_source_files_properties(can_native_posix_linux_socketcan.c
```
Note this will only affect the native_posix build of this file (i.e. when this file is build with the rest of Zephyr).
Though as things build fine now for native_sim (and also for native_posix if these are removed), I guess the `_BSD_SOURCE` & `_DEFAULT_SOURCE` are not really needed?
Anyhow, maybe you want to move this inside the `if (CONFIG_NATIVE_APPLICATION)`

(Sidenote: These are the default options used in native_sim `-D_POSIX_C_SOURCE=200809 -D_XOPEN_SOURCE=600 -D_XOPEN_SOURCE_EXTENDED`)

henrikbrixandersen replied:
Perhaps they are indeed not needed, but the Linux headers are a nightmare in this regard. I've left them in, but moved them to inside the `if` clause. Thanks.

aescolar reviewed (commented):
Thanks @henrikbrixandersen :)
Just some minors

* About `linux_socketcan_setsockopt()` and `linux_socketcan_getsockopt()`: They don't seem to be used by anything(?). But they use a socklen_t type which in principle could be different between host and Zephyr(?). Maybe the easiest thing is just to remove them.

* May I ask for the PR to also update this line
https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/posix/native_sim/doc/index.rst?plain=1#L608
to say "all"

henrikbrixandersen committed:
drivers: can: native_posix_linux: remove unused functions

Remove the functions linux_socketcan_setsockopt() and
linux_socketcan_getsockopt() as they are unused.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen commented:
Thank you for the swift review and feedback.

> * About `linux_socketcan_setsockopt()` and `linux_socketcan_getsockopt()`: They don't seem to be used by anything(?). But they use a socklen_t type which in principle could be different between host and Zephyr(?). Maybe the easiest thing is just to remove them.

I hadn't noticed that, thank. I have removed the two functions.

> * May I ask for the PR to also update this line
>   https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/posix/native_sim/doc/index.rst?plain=1#L608
>   to say "all"

Updated, thanks.

martinjaeger reviewed (approved):
Looks good. Thank you and sorry for the delayed review.

henrikbrixandersen commented:
> Looks good. Thank you and sorry for the delayed review.

No worries, thanks for reviewing.
